### PR TITLE
Recording rule for Gluster Volume Used Bytes

### PR DESCRIPTION
**Pull request by umangachapagain**
Fixes : /issues/15
Signed-off-by: Umanga Chapagain <chapagainumanga@gmail.com>

### CODE REVIEW

File: rules/gluster-volume.libsonnet [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  prometheusRules+:: {
+    groups+: [
+      {
+        name: 'gluster-volume.rules',
+        rules: [
+          {
+            expr: |||
+              'sum(max(gluster_subvol_capacity_used_bytes{%(glusterExporterSelector)s})) BY (volume, subvolume) BY (volume)',
```
Comment (cloudbehl) [inline rules/gluster-volume.libsonnet]: is it tested?

Comment (cloudbehl) [reply rules/gluster-volume.libsonnet]: @umangachapagain can you test it on a standalone Prometheus. Just the Expression and see the value expected is correct. Screenshot of how it looks would be great.

**Comment by umangachapagain** [comment]
@shtripat @cloudbehl Made final changes. It passes Promtool tests. If query logic is fine then I think its done.

**Comment by umangachapagain** [comment]
> @umangachapagain tested?
> 
> @nthomas-redhat @shtripat Make sure this gets merged before #14 PR

Had few issues. Solved it and should work fine. 
Will send in a PR asap.

**Comment by cloudbehl** [comment]
@umangachapagain we have the setup(Kube+gcs) you can apply and test there now.

Will PM you the details

**Comment by umangachapagain** [comment]
> @umangachapagain we have the setup(Kube+gcs) you can apply and test there now.
> 
> Will PM you the details

Saw that. It's still not up to the mark. We need to add underlying resources as @anmol mentioned.

**Comment by cloudbehl** [comment]
> > @umangachapagain we have the setup(Kube+gcs) you can apply and test there now.
> > Will PM you the details
> 
> Saw that. It's still not up to the mark. We need to add underlying resources as anmol mentioned.

Please check it has most of the resources created already.

**Comment by umangachapagain** [comment]
volume recording rule

@shtripat  @cloudbehl

File: rules/gluster-volume.libsonnet (hunk 1 above)
Comment (umangachapagain) [reply rules/gluster-volume.libsonnet]: ack.

**Comment by cloudbehl** [comment]
Looks good, @umangachapagain please squash the commits.

**Comment by cloudbehl** [comment]
@shtripat @nthomas-redhat Please review and approve this PR.

File: rules/gluster-volume.libsonnet [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
+  prometheusRules+:: {
+    groups+: [
+      {
+        name: 'gluster-volume.rules',
+        rules: [
+          {
+            expr: |||
+               sum(max(gluster_subvol_capacity_used_bytes{%(glusterExporterSelector)s}) BY (volume, subvolume)) BY (volume)
```
Comment (anmolsachan) [inline rules/gluster-volume.libsonnet]: I know this PR is merged and I missed to review this PR. Can you please explain the logic behind doing a 'max'  here. @umangachapagain

Comment (anmolsachan) [reply rules/gluster-volume.libsonnet]: @cloudbehl @shtripat @umangachapagain Shouldn't it be just : sum(gluster_subvol_capacity_used_bytes)by(volume)

Comment (umangachapagain) [reply rules/gluster-volume.libsonnet]: http://aravindavk.in/blog/monitoring-glusterfs-volume-utilization/
This should clear your doubts.

Comment (anmolsachan) [reply rules/gluster-volume.libsonnet]: Shouldn't this be enough : ` max(gluster_subvol_capacity_used_bytes{job="glusterd2-client"})by(volume) `

Comment (umangachapagain) [reply rules/gluster-volume.libsonnet]: Consider a Volume made of 2 - replica 3 subvolumes.
To get `used bytes` of this `volume` we need to `sum used bytes` of each `subvolume` in a volume. Hence, `sum by volume`
To get `used bytes` of each `subvolume`, we need to get `max` of `used bytes` of `replica of each subvolumes` for each `volume`. Hence, `sum by subvolume and volume`

Comment (anmolsachan) [reply rules/gluster-volume.libsonnet]: I tested. Looks Good. Thanks for explaining

**Pull request merged**